src/ec-evaluator/interpreter.ts

```diff
@@ -557,69 +547,126 @@ const cmdEvaluators: { [type: string]: CmdEvaluator } = {
+      if (c.staticFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name)) {
+        variable = c.frame.getVariable(command.name);
+      // If instance field.
+      } else if (c.instanceFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name)) {
+        // Throw NullPointerException if instance field but target is null.
+        if (isNull(varOrClass.value as Literal | Object)) {
+          throw new errors.NullPointerException();
+        }
+
+        // Only declared or inherited fields of var type can be accessed.
+        const obj = varOrClass.value as Object;
+        let objFrameClass = obj.class;
+        let objFrame = obj.frame
+        while (objFrameClass.frame.name !== varOrClass.type && objFrameClass.superclass) {
+          objFrameClass = objFrameClass.superclass;
+          objFrame = objFrame.parent;
+        }
+        variable = objFrame.getVariable(command.name)
+      } else {
+        throw new errors.UndeclaredVariableError(command.name)
+      }
+    } else /*if (varOrClass.kind === StructType.CLASS)*/ {
+      variable = varOrClass.frame.getVariable(command.name);
+    }
+
+    stash.push(variable);
   },
 
   [InstrType.DEREF]: (
     command: DerefInstr,
-    context: Context,
+    environment: Environment,
     control: Control,
     stash: Stash,
   ) => {
-    const variable = stash.pop()! as Variable;
-    stash.push(variable.value as Literal | Object);
+    const varOrClass = stash.pop()! as Variable | Class;
+    const value = varOrClass.kind === StructType.CLASS
+      ? varOrClass
+      // Variable can store variable now.
+      : varOrClass.value.kind === StructType.VARIABLE
+      ? varOrClass.value.value as Literal | Object
+      : varOrClass.value as Literal | Object;
+    stash.push(value);
   },
 
   [InstrType.NEW]: (
     command: NewInstr,
-    context: Context,
+    environment: Environment,
     control: Control,
     stash: Stash,
   ) => {
     // To be restore after extending curr env to declare instance fields in obj frame.
-    const currEnv = context.environment.current;
-    context.environment.extendEnv(command.c.frame, "object");
-
-    // Declare declared and inherited instance fields.
-    let currClass: Class | undefined = command.c;
-    while (currClass) {
-      currClass.instanceFields.forEach(i => {
+    const currEnv = environment.current;
+
+    // Get class hierarchy.
+    const objClass = command.c;
+    let currClass: Class = objClass;
+    const classHierachy: Class[] = [currClass];
+    while (currClass.superclass) {
+      classHierachy.unshift(currClass.superclass);
+      currClass = currClass.superclass;
+    };
+    
+    // Create obj with both declared and inherited fields.
+    let obj: Object;
+    classHierachy.forEach((c, i) => {
+      // Only 1 object is created.
+      i === 0 ? (obj = environment.createObj(objClass)) : environment.extendEnv(environment.current, NO_FRAME_NAME);
```

Comment:
```suggestion
    let obj: Object = environment.createObj(objClass);
    classHierachy.forEach((c, i) => {
      // Only 1 object is created.
      if (i > 0) environment.extendEnv(environment.current, NO_FRAME_NAME);
```

How about this so that in the next few lines u dont need to do `obj!`